action/protocol/staking/candidate_center_extra.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2022 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+func (cb *candBase) clone() *candBase {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	clone := newCandBase()
+	for name, cand := range cb.nameMap {
+		clone.nameMap[name] = cand.Clone()
+	}
+	for owner, cand := range cb.ownerMap {
+		clone.ownerMap[owner] = cand.Clone()
+	}
+	for operator, cand := range cb.operatorMap {
+		clone.operatorMap[operator] = cand.Clone()
+	}
+	for bucket, cand := range cb.selfStkBucketMap {
+		clone.selfStkBucketMap[bucket] = cand.Clone()
+	}
+	if len(cb.owners) > 0 {
+		for _, cand := range cb.owners {
+			clone.owners = append(clone.owners, cand.Clone())
+		}
+	}
+	return clone
+}
+
+func (cb *candBase) candsInNameMap() CandidateList {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	if len(cb.nameMap) == 0 {
+		return nil
+	}
+
+	list := make(CandidateList, 0, len(cb.nameMap))
+	for _, d := range cb.nameMap {
+		list = append(list, d.Clone())
+	}
+	return list
+}
+
+func (cb *candBase) candsInOperatorMap() CandidateList {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	if len(cb.operatorMap) == 0 {
+		return nil
+	}
+
+	list := make(CandidateList, 0, len(cb.operatorMap))
+	for _, d := range cb.operatorMap {
+		list = append(list, d.Clone())
+	}
+	return list
+}
+
+func (cb *candBase) ownersList() CandidateList {
+	cb.lock.RLock()
+	defer cb.lock.RUnlock()
+	if len(cb.owners) == 0 {
+		for _, d := range cb.ownerMap {
+			cb.owners = append(cb.owners, d.Clone())
+		}
+	}
+	return cb.owners
+}
+
+func (cb *candBase) recordOwner(c *Candidate) {
+	cb.lock.Lock()
+	defer cb.lock.Unlock()
+	if len(cb.owners) == 0 {
+		for _, d := range cb.ownerMap {
+			cb.owners = append(cb.owners, d.Clone())
+		}
+	}
+	for i, d := range cb.owners {
+		if d.Owner.String() == c.Owner.String() {
+			cb.owners[i] = c.Clone()
```

Comment:
zhi: if `nameMap` and `operatorMap` could be type of map, why `owners` is a list?

Reply:
b/c outside calls `GetByName()` and `GetByOperator()`, map better suits that need. 
`owners` is only updated by candidate update/register tx, a list is easier and suffices for the use-case